Remove preprocessing functions from docs. (#18327)
scikit-learn/scikit-learn: doc/modules/preprocessing.rst

estimator unable to learn from other features correctly as expected.

> The ``preprocessing`` module provides a utility class
> :class:`StandardScaler`, which is a quick and easy way to perform this
> operation on an array-like dataset::

Reviewer: ```suggestion
The :mod:`~sklearn.preprocessing` module provides the
:class:`StandardScaler` utility class, which is a quick and
easy way to perform the following operation on an array-like
dataset::
```

  >>> X_scaled.std(axis=0)
  array([1., 1., 1.])

> The class implements the ``Transformer`` API to compute the mean and standard
> deviation on a training set so as to be
> able to later reapply the same transformation on the testing set.
> This class is hence suitable for use in the early steps of a
> :class:`~sklearn.pipeline.Pipeline`::

Reviewer: ```suggestion
This class implements the ``Transformer`` API to compute the mean and
standard deviation on a training set so as to be able to later re-apply the
same transformation on the testing set. This class is hence suitable for
use in the early steps of a :class:`~sklearn.pipeline.Pipeline`::
```

  >>> scaler.mean_
  array([1. ..., 0. ..., 0.33...])

  >>> scaler.scale_
  array([0.81..., 0.81..., 1.24...])

> 

Reviewer: I would move these lines right after fitting the `sklearn.preprocessing.StandardScaler` estimator (line 55).

Author: Ok

        >>> import numpy as np
        >>> print_options = np.get_printoptions()
        >>> np.set_printoptions(suppress=True)

>   

Reviewer: ```suggestion

```

  >>> X_scaled.std(axis=0)
  array([1., 1., 1.])

> ..      >>> print_options = np.set_printoptions(print_options)

Reviewer: ```suggestion
..    >>> print_options = np.set_printoptions(print_options)
```

..    >>> print_options = np.set_printoptions(print_options)

This class implements the ``Transformer`` API to compute the mean and
standard deviation on a training set so as to be able to later re-apply the
same transformation on the testing set. This class is hence suitable for

> use in the early steps of a :class:`~sklearn.pipeline.Pipeline`::

Author: I decided not to do it here since there are similar examples of how to use the Pipeline class in the corresponding section of the documentation, i.e. to keep this article less convoluted. Though, I can copy and paste it here if you think that will be beneficial.

> use in the early steps of a :class:`~sklearn.pipeline.Pipeline`::

Reviewer 2: Yes, I think this section should explicitly illustrate the use of a pipeline

> use in the early steps of a :class:`~sklearn.pipeline.Pipeline`::

Author: ok

passing ``with_mean=False`` or ``with_std=False`` to the constructor
of :class:`StandardScaler`.

See also :func:`scale`.

> 

Reviewer: I would prefer to see the `sklearn.preprocessing.scale` function only in the `See Also` section of `sklearn.preprocessing.StandardScaler` reference.

Author: Do you mean [here](https://scikit-learn.org/dev/modules/generated/sklearn.preprocessing.StandardScaler.html#sklearn.preprocessing.StandardScaler)?
They already have it in that documentation, so I can just delete this piece of code.

Reviewer: Yep, right there.

  array([2.,  1.,  2.])

> See also :func:`minmax_scale` and :func:`maxabs_scale`.

Reviewer: Same here, but for `sklearn.preprocessing.MinMaxScaler` and `sklearn.preprocessing.MaxAbsScaler`.

Author: Ok, same here

thus rarely is a sensible thing to do. However, it can make sense to scale
sparse inputs, especially if features are on different scales.

> :class:`MaxAbsScaler` was specifically designed
> for scaling sparse data, and is the recommended way to go about this.
> However, :class:`StandardScaler` can accept ``scipy.sparse``

Reviewer: Nit:

```suggestion
:class:`MaxAbsScaler` was specifically designed for scaling
sparse data, and is the recommended way to go about this.
However, :class:`StandardScaler` can accept ``scipy.sparse``
```

If your data contains many outliers, scaling using the mean and variance
of the data is likely to not work very well. In these cases, you can use

> :class:`RobustScaler` as a drop-in replacement instead.
> It uses more robust estimates for the center and range of your data.

Reviewer: Nit:

```suggestion
:class:`RobustScaler` as a drop-in replacement instead. It uses
more robust estimates for the center and range of your data.
```

.. topic:: Scaling a 1D array

>    The module provides convenience functions (i.e. :func:`scale`,
>    :func:`minmax_scale`, :func:`maxabs_scale`, and :func:`robust_scale`),
>    which accept 1D array and can be useful in some specific case.

Reviewer: If we decide to keep this topic (I am -1), we have to add a warning note:

```suggestion
   The module provides convenience functions (i.e. :func:`scale`,
   :func:`minmax_scale`, :func:`maxabs_scale`, and :func:`robust_scale`),
   which accept 1D array and can be useful in some specific case.
   
   .. warning:: Risk of data leak
        Do not use these functions unless you know what you are
        doing. A common mistake is to apply it to the entire data
        *before* splitting into training and test sets. This will bias
        the model evaluation because information would have leaked
        from the test set to the training set. In general, we recommend
        using the transformers within a :class:`~sklearn.pipeline.Pipeline`
        in order to prevent most risks of data leaking.
```

Author: I'll just remove altogether

Mapping to a Uniform distribution
---------------------------------

> :class:`QuantileTransformer` provides a
> non-parametric transformation to map the data to a uniform distribution
> with values between 0 and 1::

Reviewer: Nit:

```suggestion
:class:`QuantileTransformer` provides a non-parametric
transformation to map the data to a uniform distribution
with values between 0 and 1::
```

  ... # doctest: +SKIP
  array([ 0.01...,  0.25...,  0.46...,  0.60... ,  0.94...])

> 

Reviewer: Is this blank line necessary here for a proper rendering?

<https://en.wikipedia.org/wiki/Vector_Space_Model>`_ often used in text
classification and clustering contexts.

> The :mod:`~sklearn.preprocessing` module provides a utility class

Reviewer 3: We did not discuss that we could keep the `normalize` function since they are per sample and not per feature statistic and it does not really matter if one use the transformer or the function?

Reviewer: You are absolutely right @glemaitre: https://github.com/scikit-learn/scikit-learn/issues/17402#issuecomment-640095189

Reviewer 2: indeed let's not touch the Normalizer, section, neither the `Binarizer` section.

Though we'll need to update some parts like this one:

> As for the :class:`StandardScaler` and :class:`Normalizer` classes, the
preprocessing module provides a companion function :func:`binarize`

As @OlehKSS properly did already :+1:

Author: ok

<https://en.wikipedia.org/wiki/Vector_Space_Model>`_ often used in text
classification and clustering contexts.

> The ``preprocessing`` module provides a utility class
> :class:`Normalizer`, which is a quick and easy way to perform this
> operation on a single array-like dataset, either using the ``l1``,
> ``l2``, or ``max`` norms. The class implements the
> ``Transformer`` API (even though the ``fit`` method is useless in this case:
> the class is stateless as this operation treats samples independently).

Reviewer: ```suggestion
The :mod:`~sklearn.preprocessing` module provides a utility class
:class:`Normalizer`, which is a quick and easy way to perform this
operation on a single array-like dataset, either using the ``l1``,
``l2``, or ``max`` norms. This class implements the ``Transformer``
API (even though the ``fit`` method is useless in this case: the class
is stateless as this operation treats samples independently).
```

Note: L2 normalization is also known as spatial sign preprocessing.

See also :func:`normalize`.

> 

Reviewer: Same here, but for `sklearn.preprocessing.Normalizer`.

Author: ok

         [1., 0., 0.],
         [0., 0., 0.]])

> See also :func:`binarize`.

Reviewer: Same here, but for `sklearn.preprocessing.StandardScaler` and `sklearn.preprocessing.Normalizer`.

Author: o

         [1., 0., 0.],
         [0., 0., 0.]])

As for the :class:`StandardScaler` and :class:`Normalizer` classes, the
preprocessing module provides a companion function :func:`binarize`

> to be used when the transformer API is not necessary.

Reviewer 2: Let's keep this and just remove the mention to StandardScaler. it's still valuable to mention `binarize` and `normalize` since there's no risk of data leak with these ones

Author: Ok

.. topic:: Sparse input

>   :class:`Binarizer` accepts **both dense array-like

Reviewer 2: For the same reason I think we should revert this part and still mention `binarize`

Author: Ok
